File: scripts/generate-assets.js
/**
 * Generates Ginti's brand assets — app icon, adaptive icon, splash and the
 * monochrome notification icon — from a single procedurally-drawn logo mark
 * ("rising bars + spark"). Pure Node.js, no external deps.
 *
 * Run:  node scripts/generate-assets.js
 */
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

// ─── PNG encoding (RGBA, colour type 6) ──────────────────────────────────────
const CRC_TABLE = (() => {
  const t = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    t[n] = c;
  }
  return t;
})();
function crc32(buf) {
  let crc = 0xffffffff;
  for (let i = 0; i < buf.length; i++) crc = CRC_TABLE[(crc ^ buf[i]) & 0xff] ^ (crc >>> 8);
  return (crc ^ 0xffffffff) >>> 0;
}
function chunk(type, data) {
  const t = Buffer.from(type, 'ascii');
  const len = Buffer.alloc(4); len.writeUInt32BE(data.length, 0);
  const crc = Buffer.alloc(4); crc.writeUInt32BE(crc32(Buffer.concat([t, data])), 0);
  return Buffer.concat([len, t, data, crc]);
}
function writePNG(file, rgba, w, h) {
  const sig = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(w, 0); ihdr.writeUInt32BE(h, 4);
  ihdr[8] = 8; ihdr[9] = 6; // 8-bit, RGBA
  const stride = w * 4;
  const raw = Buffer.alloc((stride + 1) * h);
  for (let y = 0; y < h; y++) {
    raw[y * (stride + 1)] = 0; // filter: none
    rgba.copy ? rgba.copy(raw, y * (stride + 1) + 1, y * stride, y * stride + stride)
              : Buffer.from(rgba.buffer, y * stride, stride).copy(raw, y * (stride + 1) + 1);
  }
  const out = Buffer.concat([
    sig,
    chunk('IHDR', ihdr),
    chunk('IDAT', zlib.deflateSync(raw, { level: 9 })),
    chunk('IEND', Buffer.alloc(0)),
  ]);
  fs.writeFileSync(file, out);
}

// ─── Canvas helpers (Uint8 RGBA, src-over compositing) ───────────────────────
function blank(w, h) { return new Uint8Array(w * h * 4); }
function over(buf, w, x, y, r, g, b, a) {
  if (a <= 0 || x < 0 || y < 0 || x >= w) return;
  const i = (y * w + x) * 4;
  if (i < 0 || i + 3 >= buf.length) return;
  const sa = a / 255;
  const da = buf[i + 3] / 255;
  const oa = sa + da * (1 - sa);
  if (oa <= 0) return;
  buf[i]     = Math.round((r * sa + buf[i]     * da * (1 - sa)) / oa);
  buf[i + 1] = Math.round((g * sa + buf[i + 1] * da * (1 - sa)) / oa);
  buf[i + 2] = Math.round((b * sa + buf[i + 2] * da * (1 - sa)) / oa);
  buf[i + 3] = Math.round(oa * 255);
}
function fillRoundedRect(buf, w, h, x0, y0, x1, y1, rad, color) {
  for (let y = Math.floor(y0); y < Math.ceil(y1); y++) {
    for (let x = Math.floor(x0); x < Math.ceil(x1); x++) {
      // distance into the rounded corners
      const cx = Math.min(Math.max(x, x0 + rad), x1 - rad);
      const cy = Math.min(Math.max(y, y0 + rad), y1 - rad);
      const dx = x - cx, dy = y - cy;
      if (dx * dx + dy * dy <= rad * rad) over(buf, w, x, y, color[0], color[1], color[2], color[3]);
    }
  }
}
// 4-point concave star (astroid): sqrt(|nx|)+sqrt(|ny|) <= 1
function fillSpark(buf, w, cx, cy, R, color) {
  for (let y = Math.floor(cy - R); y <= Math.ceil(cy + R); y++) {
    for (let x = Math.floor(cx - R); x <= Math.ceil(cx + R); x++) {
      const nx = Math.abs(x - cx) / R, ny = Math.abs(y - cy) / R;
      if (Math.sqrt(nx) + Math.sqrt(ny) <= 1) over(buf, w, x, y, color[0], color[1], color[2], color[3]);
    }
  }
}

// Downsample a super-sampled canvas by averaging SS×SS blocks (premultiplied).
function downsample(src, w, h, ss) {
  const dw = w / ss, dh = h / ss;
  const dst = blank(dw, dh);
  for (let y = 0; y < dh; y++) {
    for (let x = 0; x < dw; x++) {
      let ar = 0, ag = 0, ab = 0, aa = 0;
      for (let sy = 0; sy < ss; sy++) {
        for (let sx = 0; sx < ss; sx++) {
          const i = ((y * ss + sy) * w + (x * ss + sx)) * 4;
          const a = src[i + 3] / 255;
          ar += src[i] * a; ag += src[i + 1] * a; ab += src[i + 2] * a; aa += a;
        }
      }
      const n = ss * ss, di = (y * dw + x) * 4;
      dst[di]     = aa > 0 ? Math.round(ar / aa) : 0;
      dst[di + 1] = aa > 0 ? Math.round(ag / aa) : 0;
      dst[di + 2] = aa > 0 ? Math.round(ab / aa) : 0;
      dst[di + 3] = Math.round((aa / n) * 255);
    }
  }
  return { buf: dst, w: dw, h: dh };
}

// Bilinear scale of an RGBA bitmap.
function scale(src, sw, sh, dw, dh) {
  const dst = blank(dw, dh);
  for (let y = 0; y < dh; y++) {
    for (let x = 0; x < dw; x++) {
      const gx = (x / dw) * (sw - 1), gy = (y / dh) * (sh - 1);
      const x0 = Math.floor(gx), y0 = Math.floor(gy);
      const x1 = Math.min(x0 + 1, sw - 1), y1 = Math.min(y0 + 1, sh - 1);
      const fx = gx - x0, fy = gy - y0, di = (y * dw + x) * 4;
      for (let c = 0; c < 4; c++) {
        const a = src[(y0 * sw + x0) * 4 + c], b = src[(y0 * sw + x1) * 4 + c];
        const d = src[(y1 * sw + x0) * 4 + c], e = src[(y1 * sw + x1) * 4 + c];
        dst[di + c] = Math.round(a * (1 - fx) * (1 - fy) + b * fx * (1 - fy) + d * (1 - fx) * fy + e * fx * fy);
      }
    }
  }
  return dst;
}

function compositeCentered(dst, dw, dh, src, sw, sh, scaleFactor, offsetY = 0) {
  const tw = Math.round(dw * scaleFactor);
  const th = Math.round(tw * (sh / sw));
  const scaled = scale(src, sw, sh, tw, th);
  const ox = Math.round((dw - tw) / 2), oy = Math.round((dh - th) / 2 + offsetY);
  for (let y = 0; y < th; y++) {
    for (let x = 0; x < tw; x++) {
      const i = (y * tw + x) * 4;
      over(dst, dw, ox + x, oy + y, scaled[i], scaled[i + 1], scaled[i + 2], scaled[i + 3]);
    }
  }
}

// ─── The logo mark: 3 ascending rounded bars + a spark ───────────────────────
const WHITE = [255, 255, 255, 255];
function renderMark(size, ss) {
  const W = size * ss;
  const buf = blank(W, W);
  const u = W / 1024; // design in a 1024 grid
  const barW = 150 * u, gap = 60 * u;
  const groupW = barW * 3 + gap * 2;
  const startX = (W - groupW) / 2;
  const baseline = 745 * u;
  const heights = [240 * u, 385 * u, 530 * u];
  for (let i = 0; i < 3; i++) {
    const x0 = startX + i * (barW + gap);
    const yTop = baseline - heights[i];
    fillRoundedRect(buf, W, W, x0, yTop, x0 + barW, baseline, barW / 2, WHITE);
  }
  // Spark above the tallest (rightmost) bar
  const sparkCx = startX + 2 * (barW + gap) + barW / 2 + 70 * u;
  const sparkCy = baseline - heights[2] - 70 * u;
  fillSpark(buf, W, sparkCx, sparkCy, 95 * u, WHITE);
  fillSpark(buf, W, sparkCx + 150 * u, sparkCy + 60 * u, 42 * u, WHITE); // tiny companion
  return downsample(buf, W, W, ss);
}

// Premium rounded tile: 3-stop diagonal gradient + glossy top-left highlight +
// bottom-right depth vignette. radFrac = 0 gives a full-bleed square (for Android
// adaptive foreground, which the OS masks into its own shape).
function gradientTile(size, _top, _bottom, radFrac) {
  const buf = blank(size, size);
  const rad = size * radFrac;
  const c0 = [83, 74, 226];   // deep indigo  (top-left)
  const c1 = [108, 99, 255];  // brand violet (middle)
  const c2 = [162, 150, 255]; // soft lavender (bottom-right)
  const lerp = (a, b, t) => a + (b - a) * t;
  const clamp = (v) => (v < 0 ? 0 : v > 255 ? 255 : v);

  const hx = size * 0.30, hy = size * 0.24, hr = size * 0.62; // highlight centre
  const vx = size * 0.86, vy = size * 0.9, vr = size * 0.8;   // vignette centre

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const cx = Math.min(Math.max(x, rad), size - rad);
      const cy = Math.min(Math.max(y, rad), size - rad);
      const dx = x - cx, dy = y - cy;
      if (dx * dx + dy * dy > rad * rad) continue;

      const t = (x + y) / (2 * (size - 1)); // diagonal 0..1
      let r, g, b;
      if (t < 0.5) {
        const tt = t / 0.5;
        r = lerp(c0[0], c1[0], tt); g = lerp(c0[1], c1[1], tt); b = lerp(c0[2], c1[2], tt);
      } else {
        const tt = (t - 0.5) / 0.5;
        r = lerp(c1[0], c2[0], tt); g = lerp(c1[1], c2[1], tt); b = lerp(c1[2], c2[2], tt);
      }

      // glossy highlight (additive white, soft falloff)
      const hd = Math.hypot(x - hx, y - hy) / hr;
      const hi = Math.max(0, 1 - hd);
      const glow = hi * hi * 60;
      r += glow; g += glow; b += glow;

      // depth vignette (multiplicative darken toward bottom-right)
      const vd = Math.hypot(x - vx, y - vy) / vr;
      const vig = 1 - Math.max(0, 1 - vd) * 0.22;
      r *= vig; g *= vig; b *= vig;

      const i = (y * size + x) * 4;
      buf[i] = clamp(Math.round(r)); buf[i + 1] = clamp(Math.round(g)); buf[i + 2] = clamp(Math.round(b)); buf[i + 3] = 255;
    }
  }
  return buf;
}

// ─── Build the files ─────────────────────────────────────────────────────────
const ASSETS = path.join(__dirname, '..', 'assets');
fs.mkdirSync(ASSETS, { recursive: true });

const PURPLE_TOP = [124, 116, 255];   // #7C74FF
const PURPLE_BOT = [91, 81, 232];     // #5B51E8
const DARK = [10, 10, 10];            // #0A0A0A

const SS = 4;
const mark = renderMark(1024, SS); // { buf, w, h } white-on-transparent

// 1. icon.png — full gradient tile + white mark (iOS / legacy launcher)
{
  const S = 1024;
  const canvas = gradientTile(S, PURPLE_TOP, PURPLE_BOT, 0.235);
  compositeCentered(canvas, S, S, mark.buf, mark.w, mark.h, 0.58);
  writePNG(path.join(ASSETS, 'icon.png'), canvas, S, S);
  console.log('✓ icon.png (1024×1024)');
}

// 2. adaptive-icon.png — full-bleed premium gradient + mark (Android adaptive
//    foreground). radFrac 0 = square; the OS masks it into the launcher shape.
//    Mark kept inside the ~66% safe zone so masking never clips it.
{
  const S = 1024;
  const canvas = gradientTile(S, PURPLE_TOP, PURPLE_BOT, 0);
  compositeCentered(canvas, S, S, mark.buf, mark.w, mark.h, 0.5);
  writePNG(path.join(ASSETS, 'adaptive-icon.png'), canvas, S, S);
  console.log('✓ adaptive-icon.png (1024×1024)');
}

// 3. notification-icon.png — white silhouette on transparent (Android status bar)
{
  const S = 96;
  const small = scale(mark.buf, mark.w, mark.h, S, S);
  // Force pure white where visible (status bar tints it anyway).
  for (let i = 0; i < small.length; i += 4) {
    if (small[i + 3] > 10) { small[i] = 255; small[i + 1] = 255; small[i + 2] = 255; }
  }
  writePNG(path.join(ASSETS, 'notification-icon.png'), small, S, S);
  console.log('✓ notification-icon.png (96×96)');
}

// 4. splash.png — dark canvas with a centered gradient logo tile
{
  const W = 1284, H = 2778;
  const canvas = blank(W, H);
  for (let i = 0; i < canvas.length; i += 4) {
    canvas[i] = DARK[0]; canvas[i + 1] = DARK[1]; canvas[i + 2] = DARK[2]; canvas[i + 3] = 255;
  }
  const tileSize = 460;
  const tile = gradientTile(tileSize, PURPLE_TOP, PURPLE_BOT, 0.235);
  compositeCentered(tile, tileSize, tileSize, mark.buf, mark.w, mark.h, 0.58);
  // place tile centered
  const ox = Math.round((W - tileSize) / 2), oy = Math.round((H - tileSize) / 2);
  for (let y = 0; y < tileSize; y++)
    for (let x = 0; x < tileSize; x++) {
      const i = (y * tileSize + x) * 4;
      over(canvas, W, ox + x, oy + y, tile[i], tile[i + 1], tile[i + 2], tile[i + 3]);
    }
  writePNG(path.join(ASSETS, 'splash.png'), canvas, W, H);
  console.log('✓ splash.png (1284×2778)');
}

// ─── Google Play store graphics ──────────────────────────────────────────────
const PLAY = path.join(ASSETS, 'play');
fs.mkdirSync(PLAY, { recursive: true });

// 5. play-icon-512.png — Play Console hi-res icon (512×512, full-bleed, no transparent
//    corners; Play applies its own masking/rounding).
{
  const S = 512;
  const canvas = gradientTile(S, PURPLE_TOP, PURPLE_BOT, 0);
  compositeCentered(canvas, S, S, mark.buf, mark.w, mark.h, 0.56);
  writePNG(path.join(PLAY, 'play-icon-512.png'), canvas, S, S);
  console.log('✓ play/play-icon-512.png (512×512)');
}

// 6. feature-graphic-1024x500.png — premium gradient banner with the logo mark.
{
  const W = 1024, H = 500;
  const c0 = [83, 74, 226], c1 = [108, 99, 255], c2 = [162, 150, 255];
  const lerp = (a, b, t) => a + (b - a) * t;
  const clamp = (v) => (v < 0 ? 0 : v > 255 ? 255 : v);
  const canvas = blank(W, H);
  const hx = W * 0.25, hy = H * 0.2, hr = W * 0.6;
  for (let y = 0; y < H; y++) {
    for (let x = 0; x < W; x++) {
      const t = (x / W + y / H) / 2;
      let r, g, b;
      if (t < 0.5) { const tt = t / 0.5; r = lerp(c0[0], c1[0], tt); g = lerp(c0[1], c1[1], tt); b = lerp(c0[2], c1[2], tt); }
      else { const tt = (t - 0.5) / 0.5; r = lerp(c1[0], c2[0], tt); g = lerp(c1[1], c2[1], tt); b = lerp(c1[2], c2[2], tt); }
      const hd = Math.hypot(x - hx, y - hy) / hr, hi = Math.max(0, 1 - hd);
      const glow = hi * hi * 55; r += glow; g += glow; b += glow;
      const i = (y * W + x) * 4;
      canvas[i] = clamp(Math.round(r)); canvas[i + 1] = clamp(Math.round(g)); canvas[i + 2] = clamp(Math.round(b)); canvas[i + 3] = 255;
    }
  }
  // logo mark on the left third
  const mh = 300, mw = Math.round(mh * (mark.w / mark.h));
  const scaled = scale(mark.buf, mark.w, mark.h, mw, mh);
  const ox = Math.round(W * 0.18), oy = Math.round((H - mh) / 2);
  for (let y = 0; y < mh; y++)
    for (let x = 0; x < mw; x++) {
      const i = (y * mw + x) * 4;
      over(canvas, W, ox + x, oy + y, scaled[i], scaled[i + 1], scaled[i + 2], scaled[i + 3]);
    }
  writePNG(path.join(PLAY, 'feature-graphic-1024x500.png'), canvas, W, H);
  console.log('✓ play/feature-graphic-1024x500.png (1024×500)');
}

console.log('\nAll Ginti assets generated in assets/ (+ Play graphics in assets/play/)');

File: scripts/generate-assets.py
#!/usr/bin/env python3
"""
Generates Ginti's brand assets from two hand-designed source images:

  assets/brand/icon-source.png   — the app icon artwork (dark-green rounded
                                    tile with the wallet + leaf mark on a cream
                                    background)
  assets/brand/banner.png        — the wide marketing banner (also the source
                                    for the Play Store feature graphic)

Outputs (overwrites in place):
  assets/icon.png                        1024x1024  full-bleed app icon
  assets/adaptive-icon.png               1024x1024  Android adaptive foreground
  assets/notification-icon.png             96x96    white leaf silhouette (alpha)
  assets/splash.png                      1284x2778  deep-green splash w/ icon tile
  assets/play/play-icon-512.png            512x512  Play Console hi-res icon
  assets/play/feature-graphic-1024x500.png 1024x500 Play feature graphic

Requires Pillow:  pip install Pillow
Run:              python3 scripts/generate-assets.py
"""
import os
import math
from collections import deque
from PIL import Image, ImageDraw, ImageFilter

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
ASSETS = os.path.join(ROOT, "assets")
PLAY = os.path.join(ASSETS, "play")
BRAND = os.path.join(ASSETS, "brand")
ICON_SRC = os.path.join(BRAND, "icon-source.png")
BANNER_SRC = os.path.join(BRAND, "banner.png")

# Brand colours (sampled from the icon artwork) — kept in sync with app.json.
SPLASH_BG = (10, 33, 24)      # #0A2118 deep green
GRAD_TL = (34, 103, 79)       # icon field, top-left (lightest)
GRAD_BR = (10, 60, 50)        # icon field, bottom-right (darkest)

# Feature-graphic crop: bias left so the "Ginti" logo + phones are preserved
# (a centre crop would clip the logo). Fraction of the horizontal overscan.
FG_LEFT_FRAC = 0.043


def bri(p):
    return 0.299 * p[0] + 0.587 * p[1] + 0.714 * p[2]


def is_green(p):
    return p[1] > p[0] + 8 and p[1] > p[2] + 3 and bri(p) < 170


def detect_rect(im):
    """Bounding box of the dark-green rounded tile (dark pixels) in the icon art."""
    w, h = im.size
    px = im.load()
    minx, miny, maxx, maxy = w, h, -1, -1
    for y in range(0, h, 2):
        for x in range(0, w, 2):
            if bri(px[x, y]) < 135:
                minx = min(minx, x); maxx = max(maxx, x)
                miny = min(miny, y); maxy = max(maxy, y)
    return minx, miny, maxx, maxy


def build_fullbleed(rect):
    """Remove the cream corners of the rounded tile and return a seamless,
    full-bleed green square (gradient + wallet preserved)."""
    side = rect.width
    rpx = rect.load()

    # Flood the non-green regions connected to the 4 corners = the cream corners
    # (the centred wallet + sparkle islands are fenced off by green, so safe).
    flood = bytearray(side * side)
    inb = lambda x, y: 0 <= x < side and 0 <= y < side
    dq = deque()
    for sx, sy in [(0, 0), (side - 1, 0), (0, side - 1), (side - 1, side - 1)]:
        if not is_green(rpx[sx, sy]):
            flood[sy * side + sx] = 1
            dq.append((sx, sy))
    while dq:
        x, y = dq.popleft()
        for nx, ny in ((x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)):
            if inb(nx, ny) and not flood[ny * side + nx] and not is_green(rpx[nx, ny]):
                flood[ny * side + nx] = 1
                dq.append((nx, ny))

    # Fill flooded pixels by marching toward the centre until a green pixel.
    cx = cy = side / 2.0
    fb = rect.copy()
    fpx = fb.load()
    for y in range(side):
        row = y * side
        for x in range(side):
            if not flood[row + x]:
                continue
            dx, dy = cx - x, cy - y
            d = math.hypot(dx, dy)
            ux, uy = dx / d, dy / d
            for step in range(1, 360):
                gx = int(round(x + ux * step)); gy = int(round(y + uy * step))
                if inb(gx, gy) and not flood[gy * side + gx] and is_green(rpx[gx, gy]):
                    fpx[x, y] = rpx[gx, gy]
                    break

    # Smooth the filled corners (the per-pixel march leaves faint streaks):
    # blur, then composite the blur only over the dilated + feathered corners.
    mask = Image.new("L", (side, side), 0)
    mpx = mask.load()
    for y in range(side):
        for x in range(side):
            if flood[y * side + x]:
                mpx[x, y] = 255
    mask = mask.filter(ImageFilter.MaxFilter(21)).filter(ImageFilter.GaussianBlur(6))
    return Image.composite(fb.filter(ImageFilter.GaussianBlur(24)), fb, mask)


def gradient(size):
    canvas = Image.new("RGB", (size, size))
    cpx = canvas.load()
    for y in range(size):
        for x in range(size):
            t = (x + y) / (2 * (size - 1))
            cpx[x, y] = (
                int(GRAD_TL[0] + (GRAD_BR[0] - GRAD_TL[0]) * t),
                int(GRAD_TL[1] + (GRAD_BR[1] - GRAD_TL[1]) * t),
                int(GRAD_TL[2] + (GRAD_BR[2] - GRAD_TL[2]) * t),
            )
    return canvas


def main():
    os.makedirs(PLAY, exist_ok=True)
    src = Image.open(ICON_SRC).convert("RGB")
    rx0, ry0, rx1, ry1 = detect_rect(src)
    side = min(rx1 - rx0, ry1 - ry0)
    rect = src.crop((rx0, ry0, rx0 + side, ry0 + side))
    fb = build_fullbleed(rect)

    # 1. icon.png — full-bleed, opaque (iOS / legacy launcher)
    fb.resize((1024, 1024), Image.LANCZOS).save(os.path.join(ASSETS, "icon.png"))
    print("✓ icon.png (1024x1024)")

    # 2. play-icon-512.png — Play Console hi-res icon, full-bleed
    fb.resize((512, 512), Image.LANCZOS).save(os.path.join(PLAY, "play-icon-512.png"))
    print("✓ play/play-icon-512.png (512x512)")

    # 3. adaptive-icon.png — full-bleed; mark inset to ~90% so masks never clip it
    canvas = gradient(1024)
    sz = int(1024 * 0.90); off = (1024 - sz) // 2
    canvas.paste(fb.resize((sz, sz), Image.LANCZOS), (off, off))
    canvas.save(os.path.join(ASSETS, "adaptive-icon.png"))
    print("✓ adaptive-icon.png (1024x1024)")

    # 4. splash.png — deep-green canvas with a centred rounded icon tile
    W2, H2 = 1284, 2778
    sp = Image.new("RGB", (W2, H2), SPLASH_BG)
    ts = 620
    tile = fb.resize((ts, ts), Image.LANCZOS).convert("RGBA")
    tm = Image.new("L", (ts, ts), 0)
    ImageDraw.Draw(tm).rounded_rectangle([0, 0, ts - 1, ts - 1], radius=int(ts * 0.22), fill=255)
    tile.putalpha(tm)
    sp.paste(tile, ((W2 - ts) // 2, (H2 - ts) // 2), tile)
    sp.save(os.path.join(ASSETS, "splash.png"))
    print("✓ splash.png (1284x2778)")

    # 5. notification-icon.png — white leaf silhouette on transparent (Android)
    w, h = src.size
    lf = src.crop((int(w * 0.322), int(h * 0.491), int(w * 0.535), int(h * 0.683)))
    lpx = lf.load(); lw, lh = lf.size
    alpha = Image.new("L", (lw, lh), 0); apx = alpha.load()
    for y in range(lh):
        for x in range(lw):
            p = lpx[x, y]; b = bri(p)
            greenish = p[1] >= p[2] - 4 and p[1] >= p[0] - 12
            if b >= 240 or not greenish:
                apx[x, y] = 0
            elif b <= 180:
                apx[x, y] = 255
            else:
                apx[x, y] = int(255 * (240 - b) / 60)
    alpha = alpha.crop(alpha.getbbox())
    aw, ah = alpha.size
    s = 76 / max(aw, ah)
    na = alpha.resize((max(1, int(aw * s)), max(1, int(ah * s))), Image.LANCZOS)
    c96 = Image.new("RGBA", (96, 96), (255, 255, 255, 0))
    c96.paste(Image.new("RGBA", na.size, (255, 255, 255, 255)),
              ((96 - na.size[0]) // 2, (96 - na.size[1]) // 2), na)
    c96.save(os.path.join(ASSETS, "notification-icon.png"))
    print("✓ notification-icon.png (96x96)")

    # 6. feature-graphic-1024x500.png — crop-to-fill from the banner, biased left
    ban = Image.open(BANNER_SRC).convert("RGB")
    bw, bh = ban.size
    sw = round(bw * 500 / bh)
    scaled = ban.resize((sw, sh := 500), Image.LANCZOS)
    left = round((sw - 1024) * FG_LEFT_FRAC)
    scaled.crop((left, 0, left + 1024, 500)).save(
        os.path.join(PLAY, "feature-graphic-1024x500.png"))
    print("✓ play/feature-graphic-1024x500.png (1024x500)")

    print("\nAll Ginti assets regenerated from assets/brand/.")


if __name__ == "__main__":
    main()
